engine/execution/ingestion/engine.go

```diff
@@ -843,7 +843,7 @@ func newQueue(blockify queue.Blockify, queues *stdmap.QueuesBackdata) (*queue.Qu
 func enqueue(blockify queue.Blockify, queues *stdmap.QueuesBackdata) (*queue.Queue, bool) {
 	for _, queue := range queues.All() {
 		if queue.TryAdd(blockify) {
-			return queue, true
+			return queue, false
```

Comment:
curious why, should we return true to indicate the block was a new one?

Reply:
if it was added to an existing queue, it was not new - there are already queues where it fits

Reply:
sorry still confusing.

There are two cases `TryAdd` will return `true`:
1. it's a new block, and its parent can be found in the queue, but the block itself is not in the queue
2. it's not a new block, the block itself is found in the queue.

It makes sense to return `queue, false` for the second case, but what about the first case?

Reply:
In `1st` case I think its OK, because block is still new and we need to do rest of processing - ie request collections. It won't be executed down the lane (there won't be parent state commitment present).

Reply:
But we won't request collection if the block is not new. We [returned here](https://github.com/onflow/flow-go/blob/master/engine/execution/ingestion/engine.go#L458) before [fetching collection](https://github.com/onflow/flow-go/blob/master/engine/execution/ingestion/engine.go#L499). 

If it's a new block, and we don't fetch its collection, would would trigger it to execute? Wouldn't it get stuck?

Reply:
> But we won't request collection if the block is not new. We returned here before fetching collection.

Yes, but before the fix we always returned true and never actually terminated

> If it's a new block, and we don't fetch its collection, what would trigger it to execute? Wouldn't it get stuck?

Two cases I think - new block without enqueued parent (so it creates a new queue),
Second - if its parent is already in the queue, it will get executed soon and it will be tried then.

I think this is very important discussion here, but maybe it would be beneficial to add a test case to show what you think won't work, to be absolutely clear

Reply:
There are 3 cases in total here:

1. a new block without enqueued parent, it creates a new queue.
2. a new block whose parent can be found in an existing queue, it adds the new block to that queue
3. an existing block who can be found in an existing queue


For case 1 and 3, `TryAdd` will return `true`, and for case 2, `TryAdd` will return `false`

Your change doesn't affect case 2.

The `enqueue` method is treating case 1 and case 3 the same way, which I feel is confusing, not sure if distinguishing them would be better.

Before your change, the logic handles case 1 well, but seems for case 3, it might run into the case to execute the block again.

With your change, it handles case 3 well, but I'm not sure how it handles case 1, as I mentioned if `enqueue` returns `false` as the second value, it won't fetch collection, therefore I'm afraid the execution will halt. 

A test case would be to check if we add a second block before the first block has been executed, can the second block be executed.

Reply:
> A test case would be to check if we add a second block before the first block has been executed, can the second block be executed.

Sorry for the delay, but I added this testcase (or rather expanded the test I added to include extra blocks).
Looks like execution indeed continues

Reply:
Ah, there is also one more caveat - you are right, that in case 3 we exit method processing block early, treating it as already enqueued and not request the collection.
However, execution works because we re-requests collection after the parent has finished execution. Which is fortunate on one side, as execution continues, but on the other - its indeed a bug, because we should request collection as early as possible.